### PR TITLE
store/gcworker: migrate test-infra to testify for gc_worker_test.go #27657

### LINKED ISSUE #27657: migrate test-infra to testify for `store/gcworker`

### PR DESCRIPTION
**Pull request by penthaapatel**
### What problem does this PR solve?

Issue Number: closes #27657 

### What is changed and how it works?

What's Changed:

-   Migrate from pingcap/check to testify
-   Eliminate goroutine leak

### Release note

```release-note
None
```

### CODE REVIEW

**Review by tisonkun — COMMENTED** [review]
@penthaapatel please fix CI fails like `go fmt`.

**Comment by penthaapatel** [comment]
@tisonkun Just fixed using `go fmt`. 
`check_dev_2` CI still seems to be failing. Need some more guidance to proceed as I am a first-time contributor.

**Comment by penthaapatel** [comment]
/cc @tisonkun 
On running the package tests for `store/gcworker` with the **original** **unchanged** `gc_worker_test.go` file, the following tests were already failing on my system:
```
FAIL: gc_worker_test.go:1594: testGCWorkerSuite.TestGCLabelRules
FAIL: gc_worker_test.go:1582: testGCWorkerSuite.TestGCPlacementRules
FAIL: gc_worker_test.go:1538: testGCWorkerSuite.TestPhyscailScanLockDeadlock
FAIL: gc_worker_test.go:873: testGCWorkerSuite.TestResolveLockRangeInfine
FAIL: gc_worker_test.go:1350: testGCWorkerSuite.TestResolveLocksPhysical
OOPS: 21 passed, 5 FAILED
--- FAIL: TestT (28.04s)
FAIL
FAIL    github.com/pingcap/tidb/store/gcworker  28.618s
FAIL
```
Is this the reason why `check_dev_2` CI is unable to pass?

The only changes I have made with this PR are:
- improving the `func SetUpTest()` by using `testGCWorkerSuite` as the return value. 
- replacing `c.AssertXXX` with `testify/require` package
- adding `t.Parallel()` for imporved concurrency
- checking for goroutine leaks in `store/gcworker/main_test.go`

I can assure that no changes have been made to the logic of the original `gc_worker_test.go` code.

How should I proceed further? First time contributor here.

**Comment by penthaapatel** [comment]
/cc @tisonkun 
On running the package tests for `store/gcworker` with the **original** **unchanged** `gc_worker_test.go` file, the following tests for `gcworker` package were already failing on my system:
```
FAIL: gc_worker_test.go:1594: testGCWorkerSuite.TestGCLabelRules
FAIL: gc_worker_test.go:1582: testGCWorkerSuite.TestGCPlacementRules
FAIL: gc_worker_test.go:1538: testGCWorkerSuite.TestPhyscailScanLockDeadlock
FAIL: gc_worker_test.go:873: testGCWorkerSuite.TestResolveLockRangeInfine
FAIL: gc_worker_test.go:1350: testGCWorkerSuite.TestResolveLocksPhysical
OOPS: 21 passed, 5 FAILED
--- FAIL: TestT (28.04s)
```
Is this the reason why `check_dev_2` CI is unable to pass?

The only changes I have made with this PR are:
- improving the `func SetUpTest()` by using `s *testGCWorkerSuite` and `clean()` as the return values
- incorporating `func TearDownTest()` within `func SetUpTest()` 
- as a result of the above change, I used `func SetUpTest()` at the beginning of each test, to create a new `testGCWorkerSuite` struct, and used `defer clean()` to tear down the test
- replacing `c.AssertXXX` with `testify/require` package
- adding `t.Parallel()` for imporved concurrency
- checking for goroutine leaks in `store/gcworker/main_test.go`

I can assure that no changes have been made to the logic of the original `gc_worker_test.go` code.

How should I proceed further? First time contributor here.

**Comment by tisonkun** [comment]
@penthaapatel thanks for your quick update. I'm looking into this.

**Comment by tisonkun** [comment]
We're trying to avoid `suite` as mentioned https://github.com/pingcap/tidb/pull/26082#issuecomment-877230972 .

When you really need to batch some setup/teardown logics and state, you can take a look at https://github.com/pingcap/tidb/pull/27572 .

**Comment by tisonkun** [comment]
Let me try to push a follow up base on your change so that you can get the idea. Later you can reuse the experience on other tasks if you like.

**Review by tisonkun — APPROVED** [review]
@penthaapatel I push a follow up which you can take a look to get ideas on idiom. Sorry I don't have conversation for it since code expresses better than words in this case. You can argue or revert the changes if you have other better ideas.

**Comment by penthaapatel** [comment]
@tisonkun 
I am going through the latest changes you pushed to the code. Working with such a huge codebase for the very first time is a learning experience for me. I will try to understand the code and will try to make further contributions here.
Thank you for patiently guiding me. I will reach out to you if I have any further queries/ideas.

File: testkit/testmain/testmain.go [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+	goleak.TestingM
+	callback func(int) int
+}
+
+func (m *testingM) Run() int {
+	return m.callback(m.TestingM.Run())
+}
+
+// WrapTestingM returns a TestingM wrapped with callback on m.Run returning
+func WrapTestingM(m goleak.TestingM, callback func(int) int) *testingM {
```
Comment (disksing) [inline testkit/testmain/testmain.go:31]: Is it not used?

Comment (tisonkun) [reply testkit/testmain/testmain.go:31]: Used here https://github.com/pingcap/tidb/pull/27743/files#diff-48dd7515eec4786192bae5129ca431ce9ce97d09acf86b2ac2defd1616136022R39

Workaround `goleak.VerifyMain(m)` shortcoming on call back after `m.Run`.

**Review by disksing — APPROVED** [review]

**Review by tiancaiamao — APPROVED** [review]

**Pull request merged**